File: src/main/java/org/example/service/CountryService.java
package org.example.service;

import org.example.model.Country;
import org.example.persistence.hibernate.CountryRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CountryService {

    private final CountryRepository countryRepository;

    public CountryService(CountryRepository countryRepository) {
        this.countryRepository = countryRepository;
    }

    @Transactional(readOnly = true)
    public Optional<Country> findById(String code) {
        return countryRepository.findById(code);
    }

    @Transactional(readOnly = true)
    public List<Country> findAll() {
        return countryRepository.findAll();
    }

    public Country save(Country country) {
        return countryRepository.save(country);
    }

    public void deleteById(String id) {
        countryRepository.delete(id);
    }

}

File: src/main/kotlin/org/example/service/CountryService.kt
package org.example.service

import org.example.model.Country
import org.example.persistence.hibernate.CountryRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * `open` not needed thanks to kotlin spring plugin.
 */
@Service
@Transactional
class CountryService(private val countryRepository: CountryRepository) {

    @Transactional(readOnly = true)
    fun findById(code: String): Country? {
        return countryRepository.findById(code)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<Country> {
        return countryRepository.findAll().toList()
    }

    fun save(country: Country): Country {
        return countryRepository.save(country)
    }

    fun deleteById(id: String) {
        countryRepository.delete(id)
    }
}
